### PR TITLE
refactor: move logic to view model

### PR DESCRIPTION
**Pull request by kdvuong**
- move logic business to view model for owned list fragment
- initialize adapter only once, not for every time data changes
- use parcelable and intent to pass book around instead of SharedPreferences.

### CODE REVIEW

File: app/src/main/java/com/cmput301f20t21/bookfriends/ui/library/OwnedViewModel.java [hunk 1]
```diff
@@ -1,101 +1,104 @@
...
+            errorMessageObserver.setValue(BOOK_ERROR.FAIL_TO_EDIT_BOOK);
+        }
+    }
+
+    public void deleteBook(Book book) {
+        bookRepository.delete(book.getId())
+                .addOnSuccessListener(
+                        res -> {
+                            if (book.getImageUri() != null) {
+                                bookRepository.deleteImage(book.getId());
```
Comment (kdvuong) [inline app/src/main/java/com/cmput301f20t21/bookfriends/ui/library/OwnedViewModel.java:70]: we don't care if image delete success or not if the book is deleted from its own collection.

File: app/src/main/java/com/cmput301f20t21/bookfriends/ui/library/OwnedViewModel.java [hunk 2]
```diff
@@ -1,101 +1,104 @@
...
+
     private final AuthRepository authRepository = AuthRepository.getInstance();
     private final BookRepository bookRepository = BookRepository.getInstance();
-    
-    private MutableLiveData<List<Book>> books;
-    private MutableLiveData<Integer> updatedPosition;
-    private MutableLiveData<BOOK_ERROR> errorMessageObserver;
+
+    private MutableLiveData<List<Book>> books = new MutableLiveData<>(new ArrayList<>());
+    private List<Book> bookData = books.getValue();
```
Comment (c25vdw) [inline app/src/main/java/com/cmput301f20t21/bookfriends/ui/library/OwnedViewModel.java:25]: should we make this local var? we saved a bunch of getValue call but now it's easier to lost sync with bookData and books livedata.

File: app/src/main/java/com/cmput301f20t21/bookfriends/ui/library/OwnedListFragment.java [hunk 3]
```diff
@@ -51,26 +49,46 @@ public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup c
...
+        recyclerView.setHasFixedSize(true);
+
+        // use a linear layout manager
+        layoutManager = new LinearLayoutManager(getActivity());
+        recyclerView.setLayoutManager(layoutManager);
+
+        adapter = new OwnedListAdapter(vm.getBooks().getValue(), this::onItemClick, this::onDeleteBook);
+        recyclerView.setAdapter(adapter);
+
+        vm.getBooks().observe(getViewLifecycleOwner(), (List<Book> books) -> adapter.notifyDataSetChanged());
```
[the remark this answers is not in the document]
Comment (kdvuong) [reply app/src/main/java/com/cmput301f20t21/bookfriends/ui/library/OwnedListFragment.java:66]: list in java is reference so the adapter already has the reference of the list. We need to make sure to new set the book list in the vm to another new ArrayList() then we can just notify the adapter of the changes.

File: app/src/main/java/com/cmput301f20t21/bookfriends/ui/library/OwnedViewModel.java (hunk 2 above)
Comment (kdvuong) [reply app/src/main/java/com/cmput301f20t21/bookfriends/ui/library/OwnedViewModel.java:25]: I'm sure this is by reference as well so books.getValue() will always == bookData. We won't be out of sync but I think observe only gets called when we call books.setValue()

File: app/src/main/java/com/cmput301f20t21/bookfriends/ui/library/OwnedListFragment.java [hunk 4]
```diff
@@ -15,26 +15,24 @@
...
 import com.google.android.material.floatingactionbutton.FloatingActionButton;
 
-import java.util.ArrayList;
 import java.util.List;
 
 import static android.app.Activity.RESULT_OK;
 
 public class OwnedListFragment extends Fragment {
     public static final String BOOK_ACTION_KEY = "com.cmput301f20t21.bookfriends.BOOK_ACTION";
+    public static final String BOOK_EDIT_KEY = "BOOK_EDIT";
```
Comment (zhpeng811) [inline app/src/main/java/com/cmput301f20t21/bookfriends/ui/library/OwnedListFragment.java:30]: we should make the keys consistent and add `com.cmput301f20t21.bookfriends` as a prefix

File: app/src/main/java/com/cmput301f20t21/bookfriends/ui/add/AddEditActivity.java [hunk 5]
```diff
@@ -46,6 +44,11 @@
     private static final int IMAGE_PICK_CODE = 1000;
     private static final int PERMISSION_CODE = 1001;
 
+    public static final String NEW_BOOK_INTENT_KEY = "NEW_BOOK";
+    public static final String OLD_BOOK_INTENT_KEY = "OLD_BOOK";
+    public static final String UPDATED_BOOK_INTENT_KEY = "UPDATED_BOOK";
```
Comment (zhpeng811) [inline app/src/main/java/com/cmput301f20t21/bookfriends/ui/add/AddEditActivity.java:49]: same here, I think we should make all keys consistent by having a common prefix

File: app/src/main/java/com/cmput301f20t21/bookfriends/datatransfer/BookDataTransferHelper.java [hunk 6]
```diff
@@ -16,7 +16,7 @@
 import com.google.gson.GsonBuilder;
 
 // https://stackoverflow.com/a/18463758
-public class BookDataTransferHelper {
+    public class BookDataTransferHelper {
```
Comment (zhpeng811) [inline app/src/main/java/com/cmput301f20t21/bookfriends/datatransfer/BookDataTransferHelper.java:19]: I think we no longer need anything is in package, you want to just delete them as part of this PR?

File: app/src/main/java/com/cmput301f20t21/bookfriends/ui/library/OwnedListFragment.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (c25vdw) [reply app/src/main/java/com/cmput301f20t21/bookfriends/ui/library/OwnedListFragment.java:66]: maybe we should make all livedata final, which enforces not re-assigning.

**Review by c25vdw — APPROVED** [review]
lgtm

File: app/src/main/java/com/cmput301f20t21/bookfriends/ui/library/OwnedListFragment.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (kdvuong) [reply app/src/main/java/com/cmput301f20t21/bookfriends/ui/library/OwnedListFragment.java:66]: Oh nice idea, I think we have to not reassign bookData as well.

**Review by zhpeng811 — APPROVED** [review]
lgtm

**Pull request merged**